### DISCUSSION TITLE
DataGrid - How to color the background of cells

### DISCUSSION THREAD

**Discussion by TabakovAlex** [AvaloniaUI/Avalonia#8489] [Q&A]
Sorry if this is a duplicate of another question, but I didn't find the information I needed.
I want to colorize cells where the user has made changes.
As I understand it, cell coloring is available only with the help of styles and converters. 
Xaml code:

```
 <UserControl.Styles>    
    <Style Selector="DataGridCell.style1">
      <Setter Property="Background">
        <Binding Converter="{StaticResource ChangedConverterName}"  />
      </Setter>
    </Style>        
  </UserControl.Styles>
...
<DataGrid  Items="{Binding Signals}" AutoGenerateColumns="False" >
      <DataGrid.Columns>        
        <DataGridTextColumn Header="Short Name" Binding="{Binding ShortDesc}" CellStyleClasses="style1" />               
        ...
      </DataGrid.Columns>      
    </DataGrid>
```

Converter code:

```
public class ChangedConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is SignalDic sig )
            {
                if (sig.IsChangedShort)
                    return Brushes.GreenYellow;
                
                return Brushes.Transparent;
            }
            return BindingOperations.DoNothing;            
        }
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
```

But if the user has changed the data in the cell, redrawing does not work, because the converter is not called. Although the fields ShortDesc and IsChangedShort are being updated, they have the correct values.
It will fire if the user scrolls the DataGrid. But even so, the color is not always displayed.
How to make cell color update immediately? What am I doing wrong?
Should I apply a different way of coloring the cells?

**Comment by Mbucari** [comment]
Here's what I did.

Instead if using DataGridTextColumn, I used DataGridTemplateColumn like so:

```XAML
<DataGridTemplateColumn Header="Edit Text">
  <DataGridTemplateColumn.CellTemplate>
    <DataTemplate>
      <TextBox BorderThickness="0" Background="{Binding UserEditedBGColor}" TextWrapping="Wrap" Text="{Binding EditText, Mode=TwoWay}" />
    </DataTemplate>
  </DataGridTemplateColumn.CellTemplate>
</DataGridTemplateColumn>
```

And in my viewmodel:

```C#
private string _defaultText = "";
private string _editText;

public string EditText
{
  get => _editText; set
  {
    this.RaiseAndSetIfChanged(ref _editText, value);
    UserEditedBGColor = _editText == _defaultText ? null : Avalonia.Media.Brushes.Red;
    this.RaisePropertyChanged(nameof(UserEditedBGColor));
  }
}
public IBrush UserEditedBGColor { get; private set; }
```

When the user changes the TextBox's contents, the binding sets `EditText`, which check if the text is default and sets the appropriate background brush.

**Comment by timunie** [comment]
I think this will be solved once #7797 is merged. 

Background: The Binding seems to update the stored value only after edit ending in the normal TextBox. You should be able to use your Converter in the TextBox anyway.

**Comment by maxkatz6** [comment]
> Although the fields ShortDesc and IsChangedShort are being updated, they have the correct values.

But the model itself is not updated; how would it know?
This code would make more sense, but I haven't tried:
```
<Binding Path="IsChangedShort " Converter="{StaticResource ChangedConverterName}"  />
```


Approach from @Mbucari would also work, if you are fine with view-related logic in the view model.

Yet another alternative approach would be to use LoadingRow event and add use "row.Classes.Set" with a binding. But that's a bit more complicated and you would need to track binding disposal manually I think.

**Reply by timunie** [reply]
I really think the misunderstanding here is that a normal TextBox changes the underlaying property on euch key press. The DataGrid on the other hand doesn't do this. It waits until you exit the cell editing. That's irritating, at least it was for me the first time I hit a similar need.
